Add a only_on_screen container widget

psychon opened the pull request:
Fixes: https://github.com/awesomeWM/awesome/issues/1565
Signed-off-by: Uli Schlachter <psychon@znc.in>

Elv13 commented on lib/wibox/container/only_on_screen.lua:
```diff
@@ -0,0 +1,145 @@
+            widget:emit_signal("widget::layout_changed")
+        end
+    end
+end)
```
Is this the right place for such code? Maybe the wiboxes should always be redraw when the screen change. The DPI might also change (assuming that multi-DPI code ever get merged). The context have the screen. So if the drawin get redraw, then this code become unnecessary. Also, the taglist/tasklist widgets could eventually make the `s` argument optional and track their own screens/instances internally.

psychon replied:
Well, if the DPI does not change, then the widget context will not change and thus the wibox does not relayout/redraw everything... if I understand you correctly.
Put differently: Something still has to empty the layout caches on the widgets and that can only be done by the widget emitting the right signal.

Elv13 commented on lib/wibox/container/only_on_screen.lua:
```diff
@@ -0,0 +1,145 @@
+---------------------------------------------------------------------------
+--
```
Need doc

psychon replied:
Yup. I think I copied this from the `mirror` container, but that one does have some example code (which is not really possible here).

Elv13 commented on lib/wibox/container/only_on_screen.lua:
```diff
@@ -0,0 +1,145 @@
+--
+-- @author Uli Schlachter
+-- @copyright 2017 Uli Schlachter
+-- @classmod wibox.container.only_on_screen
```
Should this be in `awful.widget`? This is on the line between Awesome specific and generic. If we abide to the rule where "wibox is independent from Awesome", then it makes little sense there. On the other hand, the context have the screen, so the coupling is technically already present.

I am fine both way, but I think it's worth thinking about it.

psychon replied:
> "wibox is independent from Awesome"

Huh? Wibox is based on drawin which is part of awesome. It might have low number of interdependences, but it is still there.

Elv13 replied:
> It might have low number of interdependences, but it is still there.

It can be used on it's own with some minimal efforts (either from GTK+cairo or SVG like the doc tests). This is in line with your dependency work from a couple weeks back. Should `wibox` keep this "minimal interdependences" and exists on its own or should it keeps getting more dependent on Awesome and WMs concepts?

I think given it already is almost a different project, better keeping it that way.

Elv13 commented:
A simple test should be added in `test-screen-changes.lua`. Beside that and the comments above 👍

psychon commented:
> A simple test should be added in test-screen-changes.lua.

Where I write a widget which keeps track on which screens it was drawn and use that for `assert`s? Or how else would I verify that The Right Thing (tm) happens?

Elv13 commented:
> Or how else would I verify that The Right Thing (tm) happens?

By adding a custom property on the widget and mining the hierarchies to find it, then checks the size on that hierarchy element?

Not exactly easy[1], but it could be added to an utility module and re-used elsewhere.

Untested

```lua
 local function find_widget(h, prop)
     local w = h:get_widget()
     if widget._private and widget._private[prop] then return h end

     for _, child in ipairs(h:get_children()) do
         return find_widget(child, prop)
     end
 end
```

psychon commented:
> By adding a custom property on the widget and mining the hierarchies to find it,

Well, there is no public function to get the hierarchy. However, I just remembered `:find_widgets(x,y)` which might be enough for this...

The next thing would be: What to test? Setting this to `2` and `screen[2]` would work, but that's not all-that-dynamic. Would that be enough?

Elv13 commented:
> Well, there is no public function to get the hierarchy.

But there is a private one.

> The next thing would be: What to test? Setting this to 2 and screen[2] would work, but that's not all-that-dynamic. Would that be enough?

trowing ideas:

 * if it work when having 2 screens and only want the widget on one
 * does it works when the screens are swapped
 * does it work when the primary screen changes
 * does it works by screen index
 * does it works by screen name

psychon commented:
> if it work when having 2 screens and only want the widget on one

Added

> does it works when the screens are swapped

Huh? As in "screen's position changed"? That should not influence the test.
However, the test that I added tests both "only on screen 1" and "only on screen 2".

> does it work when the primary screen changes

The primary screen cannot be set from Lua. I could only try removing the current primary screen. Should this be added?

> does it works by screen index

Added.

> does it works by screen name

Not possible. Screen names come from RANDR, but cannot be set via `fake_add`.

blueyed commented:
From the coverage diff it looks like not everything is tested actually?!
https://codecov.io/gh/awesomeWM/awesome/compare/17fe135a9f3f94d70395730d8c3e95b9e8ffe688...5e82d754d6eda80ab828618967fba98ec7551232/diff

psychon commented:
Enhanced the test a bit more and fixed a bug in the C code that no one with a real X11 server would ever run into.

> From the coverage diff it looks like not everything is tested actually?!

I cannot test RandR output names without RandR, so the `property::outputs`-signal-stuff is not tested. Also, I'm not going to write tests for `get_children` and `set_children`. Besides that, everything should be covered now (but Travis is still building, so the results are not yet in).

psychon commented:
Coverage looks good. I will now do some squash, rebase and a force-push.

blueyed commented:
Awesome! :)
